tests/test_cache.py

```diff
@@ -0,0 +1,272 @@
+        )
+        session_2 = CachedSession(file_cache_directory=cache_dir_2)
+
+        resp_1 = session_1.get("https://now.httpbin.org")
+
+        self.assertEqual(os.path.isdir(cache_dir_1), True)
+
+        resp_2 = session_2.get("https://now.httpbin.org")
+
+        self.assertEqual(os.path.isdir(cache_dir_2), True)
+        self.assertNotEqual(resp_1.text, resp_2.text)
+
+        shutil.rmtree(cache_dir_2)
+
+        self.assertEqual(os.path.isdir(cache_dir_2), False)
+
+        resp_3 = session_2.get("https://now.httpbin.org")
+
+        self.assertEqual(os.path.isdir(cache_dir_2), True)
+        self.assertNotEqual(resp_2.text, resp_3.text)
+
+        session_3 = CachedSession(
+            file_cache_directory=cache_dir_1, fallback_cache_duration=2000
+        )
+
+        resp_4 = session_3.get("https://now.httpbin.org")
+
+        self.assertEqual(resp_1.text, resp_4.text)
+
+        shutil.rmtree(cache_dir_1)
+        shutil.rmtree(cache_dir_2)
+
+    @httpretty.activate
+    @patch("redis.Redis", mock_redis_singletons.mock_redis_client_singleton)
+    def test_redis_cache(self):
+        class FakeConnectionPool:
+            def __init__(self, name):
+                self.name = name
+
+        # our mock will be called here. Passing a connection_pool with a name
+        # makes sure that we can identify the different redis mocks
+
+        redis_mock_1 = redis.Redis(
+            connection_pool=FakeConnectionPool(name="test1")
+        )
+        redis_mock_2 = redis.Redis(
+            connection_pool=FakeConnectionPool(name="test2")
+        )
+
+        self.assertNotEqual(redis_mock_1, redis_mock_2)
+
+        def request_callback(request, uri, response_headers):
+            return [200, response_headers, json.dumps({"epoch": time.time()})]
+
+        httpretty.register_uri(
+            httpretty.GET, "https://now.httpbin.org", body=request_callback
+        )
+
+        with freeze_time("2012-01-14 12:00:01") as freezer:
+            session_1 = CachedSession(
+                redis_connection=redis_mock_1, fallback_cache_duration=500
+            )
+            session_2 = CachedSession(
+                redis_connection=redis_mock_2, fallback_cache_duration=1
+            )
+
+            resp_1 = session_1.get("https://now.httpbin.org")
+            resp_2 = session_2.get("https://now.httpbin.org")
+
+            self.assertNotEqual(resp_1.text, resp_2.text)
+
+            freezer.tick()
+
+            resp_3 = session_2.get("https://now.httpbin.org")
+
+            self.assertNotEqual(resp_2.text, resp_3.text)
+
+            session_3 = CachedSession(
+                redis_connection=redis_mock_1, fallback_cache_duration=1
+            )
+
+            resp_4 = session_3.get("https://now.httpbin.org")
+
+            self.assertEqual(resp_1.text, resp_4.text)
+
+    def test_timeout_adapter(self):
+        session = CachedSession(
+            timeout=1.2, file_cache_directory=file_cache_directory
```

Comment:
In every other way, this now looks perfect. But unfortunately I'm still experiencing intermittent failures with this test.

I know it sucks, but could you just make this 2 seconds, and then use `delay/3` and `delay/1` for the test below? The values for me have to be this high to avoid intermittent failures.

If we still see issues down the line, we might just have to disable this test until we find a better way to do it.

Reply:
Have updated it  and I created an issue at https://github.com/canonical-webteam/http/issues/9 . 

Maybe I can label it with **Help wanted**?